MoreLinq.Test/MemoizeTest.cs

```diff
@@ -0,0 +1,193 @@
+            {
+                flow.Add(i.ToString());
+
+                if (i == 3 || i == 7)
+                {
+                    //consume 1-3 already cached
+                    //add 4-5 to cache (so go to outer loop)
+                    //consume 1-7 already cached
+                    //add 8-10 to cache (so go to outer loop)
+
+                    flow.Add("enter inner loop");
+                    foreach (var j in source)
+                    {
+                        flow.Add(j.ToString());
+
+                        if (!firstVisitAtInnerLoopDone && j == 5)
+                        {
+                            firstVisitAtInnerLoopDone = true;
+                            break;
+                        }
+                    }
+                    flow.Add("exit inner loop");
+                }
+            }
+            flow.Add("exit outer loop");
+
+            flow.Add("enter last loop");
+            //consume 1-10 (all item were already cached)
+            foreach (var k in source)
+            {
+                flow.Add(k.ToString());
+            }
+            flow.Add("exit last loop");
+
+            return flow;
+        }
+
+        [Test]
+        public void MemoizeWithPartialIterationBeforeCompleteIteration()
+        {
+            var buffer = Enumerable.Range(0, 10).Memoize();
+
+            buffer.Take(5).AssertSequenceEqual(Enumerable.Range(0, 5));
+            buffer.AssertSequenceEqual(Enumerable.Range(0, 10));
+        }
+
+        [Test]
+        public void MemoizeIsLazy()
+        {
+            new BreakingSequence<int>().Memoize();
+        }
+
+        [Test]
+        public void MemoizeWithInMemoryCollection()
+        {
+            var list = Enumerable.Range(1, 10).ToList();
+
+            Assert.IsInstanceOf<List<int>>(list.Memoize());
+        }
+
+        [Test]
+        public void MemoizeEnumeratesOnlyOnce()
+        {
+            var memoized = Enumerable.Range(1, 10).AsTestingSequence().Memoize();
+
+            Assert.IsTrue(memoized.ToList().Count == 10);
+            Assert.IsTrue(memoized.ToList().Count == 10);
+        }
+
+        [Test]
+        public void MemoizeDoesNotDisponseOnEarlyExitByDefault()
+        {
+            Assert.Throws<AssertionException>(() =>
+            {
+                using (var xs = new[] { 1, 2 }.AsTestingSequence())
+                {
+                    xs.Memoize().Take(1).Consume();
+                    xs.Memoize(false).Take(1).Consume();
+                }
+            });
+        }
+
+        [Test]
+        public void MemoizeWithDisponseOnEarlyExitTrue()
+        {
+            using (var xs = new[] { 1, 2 }.AsTestingSequence())
+                xs.Memoize(true).Take(1).Consume();
+        }
+
+        [Test]
+        public void MemoizeDisponsesAfterSourceIsIteratedEntirely()
+        {
+            using (var xs = new[] { 1, 2 }.AsTestingSequence())
+                xs.Memoize().Consume();
+        }
+
+        [Test]
```

Comment:
Make this [test explicit](https://github.com/nunit/docs/wiki/Explicit-Attribute) via `[Test, Explicit]` as it _alone_ doubles the total test times according to my measurements:

| Framework | Branch  | Debug  | Factor | Release | Factor | Total  | Factor | Tests | 
|-----------|---------|-------:|-------:|--------:|-------:|-------:|-------:|------:| 
| net451    | master  | 8.499  | 1.0    | 7.801   | 1.0    | 16.300 | 1.0    | 1099  | 
| net451    | memoize | 13.841 | 1.6    | 13.377  | 1.7    | 27.218 | 1.7    | 1112  | 
| netcore10 | master  | 3.901  | 1.0    | 2.735   | 1.0    | 6.636  | 1.0    | 1072  | 
| netcore10 | memoize | 10.121 | 2.6    | 9.025   | 3.3    | 19.146 | 2.9    | 1085  |